### PR TITLE
237 batch select

### LINKED ISSUE #237: Make it possible to select multiple instructor/student files needed at once

### PR DESCRIPTION
**Pull request by neiljohari**
In this PR we address #237 to allow instructors to select required files for test suites in batch.

During our design phase, we decided that implementing this functionality on top of the existing dropdown typeahead system would result in a clunky UI.

Instead, our alternative is to present a batch selection mode that complements the typeahead instead of replaces it. The existing typeahead remains identical, except now instructors have a `Batch Select` button to the right of the typeahead:

Screen Shot 2020-12-13 at 7 33 35 PM

Clicking this button opens up a modified modal component that provides a filter & select mechanism for batch mode.

Screen Shot 2020-12-13 at 7 35 33 PM

The darkened cards represent those required by the suite, while the lighter ones represent all the options the user has. 

After the user confirms their selection, the required files are updated to represent the user's choices from the GUI.

The way this was accomplished was by introducing a unified search and select modal UI that could be launched by either the batch select for instructor files or the batch select for student files. To launch the UI, `start_batch_selection_mode` should be called. When the user presses confirm, `end_batch_selection_mode` is called. The cancel button does not result in the user losing their batch selection, it simply hides the UI.

This PR passes all relevant tests, and new tests are added to ensure batch selection functionality. We meet the global coverage requirements in the Travis build.

Fixes #237

### CODE REVIEW

**Comment by james-perretta** [comment]
Nice UI design! Here are some minor changes to make:
- Card colors: The dark blue for selected files is too dark for my tastes. Can we instead keep `gray-blue-1` as the color for all the cards, but then set opacity to somewhere around .5 for unselected cards (and back to 1 for selected cards). You could also make the text bold for selected cards if you like.
- Why `overflow: visible` for `.selected` (inside `batch-select-card`)? Could you maybe post a screenshot of why that's necessary?
- In `batch-select-card` change `cursor: grabbing` to `cursor: pointer` (the former makes me think drag and drop)
- Button colors in batch select modal: blue for "confirm", white for "cancel"
- Filtering in batch select modal seems to be case-sensitive (we want to ignore case). Also add a test case for this please. : )

There's one other trickier thing to discuss: The "batch mode" system you've implemented is a bit of a "Swiss-Army Knife" (see [pg. 18 of this document](http://umich.edu/~eecs381/handouts/C++_Coding_Standards.pdf) for a good definition of this). The ideal solution here would be to refactor the batch select functionality into a component that can `v-model` on the array of files needed. Given time constraints, I would recommend focusing on #452 first, then perhaps coming back to this after your exam.

**Comment by caywagner** [comment]
All style changes are made! We also changed the filter behavior and have tests to cover those. 

We originally had the `overflow: visible` for `.selected` items as a way to allow users to see the full file name. See the difference below.  `overflow: ellipsis` cuts off the file extension which users might need to see to decide if they want to select it or not. 

image
image

We went ahead and kept `overflow: ellipsis` for all the `batch-select-card`s for simplicity's sake. Ideally, we'd always have the file extension showing and just abbreviate the file name with an ellipsis but we didn't find an easy way of doing that with CSS.

**Comment by caywagner** [comment]
As far the "Swiss-Army Knife" aspect goes, we definitely agree. This implementation ended up being pretty clunky with how it depends on the `InstructorFile` and `SingleStudentFile` enum and the UX clashes a little bit too with the dropdown list and batch select button competing for attention. Overall, this is just us being unfamiliar with Vue and not implementing with best practices. 

If you're not happy with this code, we totally understand the reasoning not to accept this PR. At the very least, it was an interesting case study in to how the cards and dropdown lists differ in functionality and usability.

**Comment by james-perretta** [comment]
@caywagner Thanks for those style changes! There are some tweaks to the test cases we'll want to make, but let's hold off on that for now. 

`overflow: ellipsis` is good--it comes in handy in the case where there aren't any "work break" characters (spaces, hyphens) in the filename. We can easily enough tweak the wrapping behavior later on.

As for the state of this PR, I don't see any reason not to keep building on it. I'm very happy with the design work you've done so far, and we only need a bit of refactoring to fix the swiss-army knife issue. 

I'll follow up in a few minutes with a code skeleton for a "batch select" component, and then we can proceed with the refactoring from there.

**Comment by james-perretta** [comment]
Here's the boilerplate for a batch select component (`src/components/batch_select.vue`):
```src/components/batch_select.vue
<template>
  <div class="batch-select">
    <!-- include the "batch select" button, modal, and contents here -->
  </div>
</template>

<script lang="ts">
import { Component, Prop, Vue, Watch } from 'vue-property-decorator';

@Component({})
export default class BatchSelect extends Vue {
  @Prop({required: true, type: Array})
  choices!: unknown[];

  // The items already selected in the parent component.
  // For a component to support v-model, it should have an input property called "value"
  // and emit an "input" event
  @Prop({type: Array})
  value!: unknown[];

  // Equivalent to "d_batch_needed_files"
  d_selected_items: unknown[] = [];

  created() {
    this.d_selected_items = this.value.slice();
  }

  @Watch('value', {deep: true})
  on_value_change(new_value: unknown[], old_value: unknown[]) {
    this.d_selected_items = this.value.slice();
  }

  confirm_selection() {
    // Call when user clicks the "confirm" button. Analogous to "end_batch_selection_mode"
  }
}
</script>

<style scoped lang="scss">

</style>
```

And then usage in `suite_settings.vue` would be something like:
```vue
<batch-select v-model="suite.instructor_files_needed" :choices="project.instructor_files">
```

Let me know if there's anything I can clarify about this. : )

**Comment by james-perretta** [comment]
P.S. If you end up not having more time to work on this, let me know and I'll transfer your work to an issue branch in the main repo and build on that. No rush, just let me know what you decide at some point.

**Comment by sauleanf** [comment]
I could try to pick this up if that is okay with @neiljohari, @caywagner, and @james-perretta

**Comment by caywagner** [comment]
I put a couple of hours in to this yesterday and one today. As I refactor, I'm creating a lot of weird bugs that take me a while to resolve. This doesn't seem hard in theory but my unfamiliarity with Vue is slowing me down. I don't want to sink too much more of my break in to it so I wouldn't mind handing it off to @sauleanf. 

You'll need permissions from @neiljohari to push but that should be it.

**Comment by james-perretta** [comment]
@caywagner Fair enough--thanks again for your work on this, I really appreciate it!

@sauleanf Yes, you're welcome to work on this.

**Comment by sauleanf** [comment]
Sorry, I have been busy and I have only been able to take a look at it tonight. I think that I refactored it such that most of the logic was abstracted into a new batch-select component in an attempt to fix the swiss-army knife design issue. 

I will admit that I am not terribly familiar with vue.js and relied on my knowledge of react to accomplish it, hopefully without introducing any anti-patterns. I just want to verify that I am allowed to push to the forked repository before I introduce my changes.

**Comment by neiljohari** [comment]
@sauleanf I added you to my fork if you want to push to this PR

File: src/components/batch_select.vue
Comment (sauleanf) [inline src/components/batch_select.vue]: Since the component is agnostic of the data that is being passed in, we need a way to compare two unknowns (for `batch_toggle_select`) and to get name from them (for filtering in `batch_filtered_items`). Functions seem the way to go if we are assuming nothing about the items. 

However, if we assume that the items are objects with valid string and value pairs, then maybe we can pass in strings corresponding to the keys of the object that give us the unique id and name. For example, we can create a prop called `name` and pass in "pattern" so that we can have something like `item[this.name]` and that would work for student files.

File: tests/test_project_admin/test_suite_settings.ts
Comment (sauleanf) [inline tests/test_project_admin/test_suite_settings.ts:412]: Note that I have not refactored the tests as I wasn't entirely sure that my approach was what @james-perretta had in mind. I didn't want to do anything prematurely.

File: src/components/batch_select.vue
Comment (james-perretta) [reply src/components/batch_select.vue]: That's more or less the direction we'll want to go--I'll take a closer look on Monday and follow up then. Thanks!

File: src/components/batch_select.vue
Comment (james-perretta) [inline src/components/batch_select.vue]: We should be able to use a scoped slot here so that this component is more flexible. Something like: 

```vue
<slot v-bind:item="item"></slot>
```

And then in `suite_settings.vue` something like:
```vue
<batch-select
  v-model="suite.instructor_files_needed"
  :choices="project.instructor_files"
  :are_items_equal="are_files_equal"
  :filter_fn="instructor_file_filter_fn"
  @items_selected="set_instructor_files($event)"
  v-slot:default="{item}"  <!-- NEW -->
>
  {{item.name}}  <!-- use item.name for instructor file, item.pattern for expected student file -->
</batch-select>
```

File: src/components/project_admin/suite_settings.vue
Comment (james-perretta) [inline src/components/project_admin/suite_settings.vue]: It looks like there are a lot of formatting-only changes to this file, which makes it harder for me to tell what actually changed in the file. Could you please revert the formatting changes?

File: src/components/batch_select.vue
Comment (james-perretta) [inline src/components/batch_select.vue:76]: Yes, this approach is what we want.

**Review by james-perretta — COMMENTED** [review]
Looking good overall! Left a few comments, and you can go ahead with updating the tests.

File: src/components/project_admin/suite_settings.vue
Comment (sauleanf) [reply src/components/project_admin/suite_settings.vue]: It looks like that specific formatting change was done in this [commit](https://github.com/eecs-autograder/ag-website-vue/pull/453/commits/860cf62c5081a92b3bd74a81bace0e22b449fa42) as part of a lot styling changes. Should I drop that commit via a rebase? Or should I just remove the new format changes for jsx and keep the rest in a new commit?

File: src/components/batch_select.vue
Comment (sauleanf) [reply src/components/batch_select.vue]: Sorry I missed that earlier

File: tests/test_project_admin/test_suite_settings.ts
Comment (sauleanf) [reply tests/test_project_admin/test_suite_settings.ts:412]: I mostly left these tests here to verify that the props are being passed in properly.

File: src/components/project_admin/suite_settings.vue
Comment (james-perretta) [reply src/components/project_admin/suite_settings.vue]: Ah that makes sense--I had made a similar comment on Neil and Cay's other PR. If you're comfortable enough with the git wizardry to easily drop the commit, please do, otherwise I can deal with it when I make my next review pass.

File: tests/test_batch_select.ts
Comment (james-perretta) [inline tests/test_batch_select.ts]: Could you add a couple tests that cover:
- Opening and closing the modal (clicking the "Cancel" button and making the modal component emit a "close" event; I can point you towards some similar tests that do this if you like)
- The `value` array changing from outside (to test the watcher)

**Comment by sauleanf** [comment]
Would you mind if at the end, I squash all the commits in this PR into one commit?

**Comment by james-perretta** [comment]
I can squash when I merge the PR. Until then I'd like to keep the history around.

File: tests/test_batch_select.ts
Comment (james-perretta) [inline tests/test_batch_select.ts:97]: Are we missing an `expect(batch_select_wrapper.vm.d_show_batch_select_modal).toBe(false);` at the end here?

Comment (sauleanf) [reply tests/test_batch_select.ts:97]: Fixed

File: src/components/batch_select.vue
Comment (james-perretta) [inline src/components/batch_select.vue]: Sorry I didn't catch this earlier: this event should be called "input" so that the component properly supports v-model. See https://vuejs.org/v2/guide/components.html#Using-v-model-on-Components

Then, the calls to `set_student_files` and `set_instructor_files` can be replaced with `$emit("field_change", d_suite)`. I'll mark those specific places just do be clear.

File: src/components/project_admin/suite_settings.vue
Comment (james-perretta) [inline src/components/project_admin/suite_settings.vue]: Change to `@input="$emit('field_change', d_suite)"`

File: src/components/project_admin/suite_settings.vue
Comment (james-perretta) [inline src/components/project_admin/suite_settings.vue]: Change to `@input="$emit('field_change', d_suite)"`

File: src/components/project_admin/suite_settings.vue
Comment (james-perretta) [inline src/components/project_admin/suite_settings.vue]: Remove

File: src/components/project_admin/suite_settings.vue
Comment (james-perretta) [inline src/components/project_admin/suite_settings.vue]: Remove

File: src/components/batch_select.vue
Comment (sauleanf) [reply src/components/batch_select.vue]: Fixed

File: tests/test_batch_select.ts
Comment (james-perretta) [reply tests/test_batch_select.ts:97]: For some reason I don't see the added `expect` anymore... Am I missing something?

**Review by james-perretta — DISMISSED** [review]
Looks good! I'll merge once the CI checks pass

**Comment by james-perretta** [comment]
Spoke too soon--can you rebase on top of develop?

**Review by james-perretta — DISMISSED** [review]

**Comment by sauleanf** [comment]
My bad, I realized that I first rebased on the wrong develop branch so I had to drop the bad commit.

**Review by james-perretta — APPROVED** [review]

**Comment by james-perretta** [comment]
@sauleanf @caywagner @neiljohari Thanks again for all your work on this!

**Pull request merged**
